test/unit/GraphQLClient.test.js

```diff
@@ -1,5 +1,293 @@
+      expect(groupCollapsedSpy).toHaveBeenCalledWith('HTTP ERROR:');
+      expect(logSpy).toHaveBeenCalledWith('on no http!');
+      expect(groupEndSpy).toHaveBeenCalled();
+    });
+
+    it('logs all graphQLErrors', () => {
+      const client = new GraphQLClient({ ...validConfig });
+      const graphQLErrors = ['on no GraphQL!', 'oops GraphQL!'];
+      client.logErrorResult({ result: { graphQLErrors } });
+      expect(groupCollapsedSpy).toHaveBeenCalledWith('GRAPHQL ERROR:');
+      expect(logSpy).toHaveBeenCalledWith('on no GraphQL!');
+      expect(logSpy).toHaveBeenCalledWith('oops GraphQL!');
+      expect(groupEndSpy).toHaveBeenCalled();
+    });
+  });
+
+  describe('generateResult', () => {
+    it('shows as errored if there are graphQL errors', () => {
+      const client = new GraphQLClient({ ...validConfig });
+      const result = client.generateResult({
+        graphQLErrors: ['error 1', 'error 2']
+      });
+      expect(result.error).toBe(true);
+    });
+
+    it('shows as errored if there is a fetch error', () => {
+      const client = new GraphQLClient({ ...validConfig });
+      const result = client.generateResult({
+        fetchError: 'fetch error'
+      });
+      expect(result.error).toBe(true);
+    });
+
+    it('shows as errored if there is an http error', () => {
+      const client = new GraphQLClient({ ...validConfig });
+      const result = client.generateResult({
+        httpError: 'http error'
+      });
+      expect(result.error).toBe(true);
+    });
+
+    it('returns the errors & data', () => {
+      const client = new GraphQLClient({ ...validConfig });
+      const data = {
+        graphQLErrors: ['graphQL error 1', 'graphQL error 2'],
+        fetchError: 'fetch error',
+        httpError: 'http error',
+        data: 'data!'
+      };
+      const result = client.generateResult(data);
+      expect(result).toEqual({
+        error: true,
+        graphQLErrors: data.graphQLErrors,
+        fetchError: data.fetchError,
+        httpError: data.httpError,
+        data: data.data
+      });
+    });
+  });
+
+  describe('getCacheKey', () => {
+    it('returns a cache key', () => {
+      const client = new GraphQLClient({
+        ...validConfig,
+        fetchOptions: { optionOne: 1 }
+      });
+      const cacheKey = client.getCacheKey('operation', {
+        fetchOptionsOverrides: { optionTwo: 2 }
+      });
+      expect(cacheKey).toEqual({
+        operation: 'operation',
+        fetchOptions: { optionOne: 1, optionTwo: 2 }
+      });
+    });
+  });
+
+  describe('request', () => {
+    afterEach(() => {
+      fetch.resetMocks();
+    });
+
+    it('sends the request to the configured url', async () => {
+      const client = new GraphQLClient({ ...validConfig });
+      fetch.mockResponseOnce(JSON.stringify({ data: 'data' }));
+      await client.request({ query: TEST_QUERY });
+      expect(fetch.mock.calls[0][0]).toBe(validConfig.url);
```

Comment:
Minor: Sometimes it's useful to set `actual` and `expected` explicitly, making it a bit clearer to understand whats being tested:

```js
const actual = fetch.mock.calls[0][0]
const expected = validConfig.url
expect(actual).toBe(expected)
```